core/zoe/isOfferSafe.js

```diff
@@ -0,0 +1,86 @@
+import { allTrue } from './utils';
+import { insist } from '../../util/insist';
+
+/**
+ * @param  {array} assaysPerIssuer - an array of assays ordered in the
+ * same order as the corresponding issuers
+ * @param  {array} rulesPerIssuer - an array of "rules" ordered in the
+ * same order as the corresponding issuers. Rules are a player's
+ * understanding of the contract that they are entering when they make
+ * an offer. Rules are structured in the form { rule: '', amount}
+ * @param  {array} amountsPerIssuer - an array of amounts ordered in
+ * the same order as the corresponding issuers. This array of amounts
+ * is the reallocation to be given to a player.
+ *
+ * `isOfferSafeForPlayer` checks offer-safety for a single player.
+ *
+ * Note: This implementation checks whether we refund for all rules or
+ * return winnings for all rules. It does not allow some refunds and
+ * some winnings, which is what would happen if you checked the rules
+ * independently.
+ */
+function isOfferSafeForPlayer(
+  assaysPerIssuer,
+  rulesPerIssuer,
+  amountsPerIssuer,
+) {
+  insist(
+    assaysPerIssuer.length === rulesPerIssuer.length &&
+      assaysPerIssuer.length === amountsPerIssuer.length,
+  )`assays, rules, and amounts must be arrays of the same length`;
+
+  // Is the amount greater than or equal to the amount in the offer?
+  const isAmountOfferSafe = (assay, amount, rule) => {
+    return assay.includes(amount, rule.amount);
+  };
+
+  // If we are refunding the player, are their allocated amounts
+  // greater than or equal to what they said they had at the beginning?
+  const refundOk = rulesPerIssuer
+    .map((rule, i) => {
+      // If the rule was 'haveExactly', we should make sure that the
+      // user gets it back in a refund. If it is not 'haveExactly' anything
+      // we give back is fine.
+      if (rule.rule === 'haveExactly') {
+        return isAmountOfferSafe(assaysPerIssuer[i], amountsPerIssuer[i], rule);
+      }
+      return true;
+    })
+    .reduce(allTrue, true);
```

Comment:
```.reduce()``` only seems to need one arg according to test-utils.  Same issue on line 63.

Reply:
reduce takes an optional initial value, but I think you're right that `true` isn't needed here since we'll already get a good initial value from the first item.

Reply:
Removed the initial value